Kernel/FileSystem/Ext2FS/Inode.cpp

```diff
@@ -36,419 +37,317 @@ static u8 to_ext2_file_type(mode_t mode)
+    m_raw_inode.i_block[EXT2_TIND_BLOCK] = 0;
+    set_metadata_dirty(true);
+
+    return {};
 }
 
-ErrorOr<Ext2FS::BlockList> Ext2FSInode::compute_block_list() const
+ErrorOr<u32> Ext2FSInode::allocate_and_zero_block()
 {
-    return compute_block_list_impl(false);
+    auto const block_size = fs().logical_block_size();
+
+    auto blocks = TRY(fs().allocate_blocks(fs().group_index_from_inode(index()), 1));
+    auto block = blocks.first();
+
+    auto buffer_content = TRY(ByteBuffer::create_zeroed(block_size));
+    TRY(fs().write_block(block, UserOrKernelBuffer::for_kernel_buffer(buffer_content.data()), block_size));
+    return block.value();
 }
 
-ErrorOr<Ext2FS::BlockList> Ext2FSInode::compute_block_list_with_meta_blocks() const
+ErrorOr<void> Ext2FSInode::write_block_pointer(BlockBasedFileSystem::BlockIndex logical_block_index, BlockBasedFileSystem::BlockIndex on_disk_index)
 {
-    return compute_block_list_impl(true);
+    VERIFY(m_inode_lock.is_locked());
+    VERIFY(logical_block_index >= EXT2_NDIR_BLOCKS);
+
+    if (logical_block_index < singly_indirect_block_capacity())
+        return write_singly_indirect_block_pointer(logical_block_index, on_disk_index);
+
+    if (logical_block_index < doubly_indirect_block_capacity())
+        return write_doubly_indirect_block_pointer(logical_block_index, on_disk_index);
+
+    if (logical_block_index < triply_indirect_block_capacity())
+        return write_triply_indirect_block_pointer(logical_block_index, on_disk_index);
+
+    VERIFY_NOT_REACHED();
 }
 
-ErrorOr<Ext2FS::BlockList> Ext2FSInode::compute_block_list_impl(bool include_block_list_blocks) const
+ErrorOr<void> Ext2FSInode::flush_block_list(Ext2FS::BlockList const& old_block_list)
 {
-    // FIXME: This is really awkwardly factored.. foo_impl_internal :|
-    auto block_vector = TRY(compute_block_list_impl_internal(m_raw_inode, include_block_list_blocks));
+    MutexLocker locker(m_inode_lock);
+
+    for (unsigned i = 0; i < EXT2_NDIR_BLOCKS; ++i) {
+        auto block = get_block(i);
+        if (m_raw_inode.i_block[i] != block)
+            set_metadata_dirty(true);
+        m_raw_inode.i_block[i] = block.value();
+    }
 
-    Ext2FS::BlockList block_list;
-    for (size_t i = 0; i < block_vector.size(); ++i) {
-        if (block_vector[i] != 0)
-            TRY(block_list.try_set(i, block_vector[i]));
+    for (auto const& block : old_block_list) {
+        if (block.key < EXT2_NDIR_BLOCKS)
+            continue;
+        if (!m_block_list.contains(block.key))
+            TRY(write_block_pointer(block.key, 0));
     }
 
-    return block_list;
+    for (auto const& block : m_block_list) {
+        if (block.key < EXT2_NDIR_BLOCKS)
+            continue;
+        auto iterator = old_block_list.find(block.key);
+        if (iterator == old_block_list.end() || (*iterator).value != block.value)
+            TRY(write_block_pointer(block.key, block.value));
+    }
```

Comment:
please prefer structured bindings, as `block.key` and `block.value` dont mean much to people glancing at the code